# Import Modules and Data

In [1]:
# Load modules
import sys
import os
os.chdir('..')

import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings("ignore", category=FutureWarning)

import tensorflow as tf
import tensorflow.keras as keras
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Conv2D, Dense, Flatten, MaxPool2D, Dropout, BatchNormalization, Activation
from tensorflow.keras.optimizers import Adam, RMSprop
from tensorflow.keras.callbacks import ModelCheckpoint, EarlyStopping
from tensorflow.keras.preprocessing.image import ImageDataGenerator
import matplotlib.pyplot as plt
%matplotlib inline

import seaborn as sns
sns.set()

from functions import cvmodeleval,samplecv, trainsampling

In [2]:
# Set memory limit on GPU to keep it from freezing up when fitting TensorFlow models later
gpus = tf.config.experimental.list_physical_devices('GPU')
if gpus:
    # Restrict TensorFlow to only allocate 1GB * 3 of memory on the first GPU
    try:
        tf.config.experimental.set_virtual_device_configuration(gpus[0], \
                                                                [tf.config.experimental.\
                                                                 VirtualDeviceConfiguration\
                                                                 (memory_limit=1024 * 3)])
        logical_gpus = tf.config.experimental.list_logical_devices('GPU')
        print(len(gpus), "Physical GPUs,", len(logical_gpus), "Logical GPUs")
    except RuntimeError as e:
        # Virtual devices must be set before GPUs have been initialized
        print(e)

1 Physical GPUs, 1 Logical GPUs


In [3]:
# Load training filename dataframe
df = pd.read_csv('data/processed/driver_image_list_processed.csv')

In [65]:
# Labeled and unlabeled test filename dataframes
df_test_labeled = pd.read_csv('C:/Users/Dylan/Desktop/Data Science/Projects/DistractedDrivers/data/processed/labeled_test_df.csv')
df_test = pd.read_csv('C:/Users/Dylan/Desktop/Data Science/Projects/DistractedDrivers/data/raw/test_filenames.csv')

In [4]:
# Call custom function to over/undersample classes occurance by subject so dataset is completely balanced.
df = trainsampling(df, samples=80, random_state=42)

In [6]:
# Define path variables for data
train_path = 'C:/Users/Dylan/Desktop/Data Science/Projects/DistractedDrivers/data/raw/imgs/train'
labeled_test_path = 'C:/Users/Dylan/Desktop/Data Science/Projects/DistractedDrivers/data/raw/imgs/testlabeled'
unlabeled_test_path = 'C:/Users/Dylan/Desktop/Data Science/Projects/DistractedDrivers/data/raw/imgs/test'

4 subjects were chosen from the training data to be used for validation during model training. These subjects represent one woman and man with dark skin and one woman and man with light skin. This is to help balance any potential racial bias in the model.

In [7]:
# Define list of validation subjects
val_subjects = ['p056', 'p050', 'p041', 'p016']

In [8]:
# Define training and validation datasets
df_train = df[~df['subject'].isin(val_subjects)]
df_val = df[df['subject'].isin(val_subjects)]

In [9]:
# Shuffle the dataframes
df_train = df_train.sample(frac=1, random_state=42)
df_val = df_val.sample(frac=1, random_state=42)

## Pre-Processing and Data-Loading

In [ ]:
# Define ImageDataGenerators for training and test/validation data. Generators include randomized preprocessing for 
# called out parameters. Test_dgen will be used for both validation data and test data.

train_dgen = ImageDataGenerator(samplewise_center=True,
                                rescale=1./255,
                                rotation_range=40,
                                width_shift_range=0.2,
                                height_shift_range=0.2,
                                channel_shift_range=0.2,
                                shear_range=0.2,
                                zoom_range=0.2,
                                brightness_range=[0.5, 1.5])

test_dgen = ImageDataGenerator(samplewise_center=True,
                               rescale=1./255)

In [67]:
# Load training, validation, and test data
train = train_dgen.flow_from_dataframe(df_train,
                                       x_col='imgpath',
                                       y_col='classname',
                                       batch_size=16, 
                                       target_size=(227,227),
                                       shuffle=True)

val = test_dgen.flow_from_dataframe(df_val,
                                    x_col='imgpath',
                                    y_col='classname',
                                    target_size=(227,227),
                                    shuffle=False)

test_labeled = test_dgen.flow_from_dataframe(df_test_labeled,
                                             x_col='filename',
                                             y_col='classname',
                                             target_size=(227,227),
                                             shuffle=False)

test = test_dgen.flow_from_dataframe(df_test,
                                     x_col='filename',
                                     y_col='class',
                                     target_size=(227,227),
                                     shuffle=False)


Found 17600 validated image filenames belonging to 10 classes.
Found 3200 validated image filenames belonging to 10 classes.
Found 200 validated image filenames belonging to 10 classes.
Found 79726 validated image filenames belonging to 1 classes.


# Model1 - AlexNet Architecture

Architecture below is based off of AlexNet

In [10]:
# Define function for initializing model
def alexNet_arch(opt):
    model = Sequential()
    model.add(Conv2D(99,
                  kernel_size=11,
                  strides=4,
                  padding='valid',
                  input_shape=(227, 227, 3)))
    model.add(BatchNormalization())
    model.add(Activation('relu'))
    model.add(MaxPool2D(3,
                         strides=2,
                         padding='valid'))
    model.add(Conv2D(256,
                      kernel_size=5,
                      strides=1,
                      padding='same'))
    model.add(BatchNormalization())
    model.add(Activation('relu'))
    model.add(MaxPool2D(3,
                        strides=2,
                        padding='valid'))
    model.add(Conv2D(384,
                      kernel_size=3,
                      strides=1,
                      padding='same'))
    model.add(BatchNormalization())
    model.add(Activation('relu'))
    model.add(Conv2D(384,
                      kernel_size=3,
                      strides=1,
                      padding='same'))
    model.add(BatchNormalization())
    model.add(Activation('relu'))
    model.add(Conv2D(256,
                      kernel_size=3,
                      strides=1,
                      padding='same'))
    model.add(BatchNormalization())
    model.add(Activation('relu'))
    model.add(MaxPool2D(3,
                         strides=2,
                         padding='valid'))
    model.add(Flatten())
    model.add(Dense(4096, activation='relu'))
    model.add(Dropout(0.5))
    model.add(Dense(4096, activation='relu'))
    model.add(Dropout(0.5))
    model.add(Dense(10, activation='softmax'))
    model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])
    
    return model

In [11]:
# Define optimizer variable
opt = RMSprop(learning_rate=0.0001)

In [12]:
# Compile model_1
model_1 = alexNet_arch(opt)

In [15]:
# Designate ModelCheckpoint and EarlyStopping callbacks_list
checkpoint = ModelCheckpoint('C:/Users/Dylan/Desktop/Data Science/Projects/DistractedDrivers/data/weights/model1_weights.hdf5',
                              mode='min',
                              monitor='val_loss',
                              save_best_only=True)

earlystop = EarlyStopping(monitor='val_loss', min_delta=0.025, patience=20, restore_best_weights=True)

callbacks_list = [checkpoint, earlystop]

In [16]:
%%time
# Train model_1
model_1.fit(train,
            epochs=100,
            steps_per_epoch=110,
            validation_data=val,
            callbacks=callbacks_list)

  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train for 110 steps, validate for 100 steps
Epoch 1/100
110/110 [==============================] - 71s 643ms/step - loss: 4.4653 - accuracy: 0.1148 - val_loss: 2.3744 - val_accuracy: 0.1000
Epoch 2/100
110/110 [==============================] - 64s 584ms/step - loss: 2.6455 - accuracy: 0.0949 - val_loss: 2.3084 - val_accuracy: 0.1000
Epoch 3/100
110/110 [==============================] - 56s 512ms/step - loss: 2.4212 - accuracy: 0.1159 - val_loss: 2.3324 - val_accuracy: 0.1000
Epoch 4/100
110/110 [==============================] - 62s 560ms/step - loss: 2.3981 - accuracy: 0.1267 - val_loss: 2.2477 - val_accuracy: 0.1603
Epoch 5/100
110/110 [==============================] - 56s 511ms/step - loss: 2.2667 - accuracy: 0.1739 - val_loss: 2.3617 - val_accuracy: 0.1003
Epoch 6/100
110/110 [==============================] - 60s 546ms/step - loss: 2.1387 - accuracy: 0.1972 - val_loss: 2.0907 - val_accuracy: 0.2722
Epoch 7/100
110/110 [=========

110/110 [==============================] - 57s 516ms/step - loss: 0.6652 - accuracy: 0.7898 - val_loss: 2.2284 - val_accuracy: 0.6678
Epoch 56/100
110/110 [==============================] - 60s 548ms/step - loss: 0.5888 - accuracy: 0.8040 - val_loss: 1.3704 - val_accuracy: 0.6128
Epoch 57/100
110/110 [==============================] - 59s 535ms/step - loss: 0.5953 - accuracy: 0.8057 - val_loss: 1.2898 - val_accuracy: 0.5972
Epoch 58/100
110/110 [==============================] - 60s 547ms/step - loss: 0.5930 - accuracy: 0.8000 - val_loss: 1.8623 - val_accuracy: 0.5825
Epoch 59/100
110/110 [==============================] - 57s 521ms/step - loss: 0.5794 - accuracy: 0.8074 - val_loss: 1.1326 - val_accuracy: 0.6966
Epoch 60/100
110/110 [==============================] - 57s 516ms/step - loss: 0.5962 - accuracy: 0.8062 - val_loss: 3.8987 - val_accuracy: 0.4913
Epoch 61/100
110/110 [==============================] - 57s 514ms/step - loss: 0.6026 - accuracy: 0.8045 - val_loss: 1.1211 - val_a

In [17]:
# Save model predictions of validation data
validation_predictions = model_1.predict(val)

In [18]:
# Convert to a dataframe with original labeles
validation_predictions = pd.DataFrame(validation_predictions, 
                                         columns=['c0', 'c1', 'c2', 'c3', 'c4', 'c5', 'c6', 'c7', 'c8', 'c9'])
validation_predictions = df_val.reset_index(drop=True).join(validation_predictions)

subject classname            img  \
0       p050        c9  img_37057.jpg   
1       p056        c1  img_70214.jpg   
2       p050        c7  img_89317.jpg   
3       p041        c1   img_9714.jpg   
4       p016        c2  img_46158.jpg   
...      ...       ...            ...   
3195    p041        c3  img_78650.jpg   
3196    p041        c4  img_83455.jpg   
3197    p041        c6  img_58453.jpg   
3198    p041        c0  img_51874.jpg   
3199    p056        c9  img_90247.jpg   

                                                imgpath            c0  \
0     C:/Users/Dylan/Desktop/Data Science/Projects/D...  6.714016e-02   
1     C:/Users/Dylan/Desktop/Data Science/Projects/D...  9.916137e-06   
2     C:/Users/Dylan/Desktop/Data Science/Projects/D...  4.676285e-03   
3     C:/Users/Dylan/Desktop/Data Science/Projects/D...  2.411511e-07   
4     C:/Users/Dylan/Desktop/Data Science/Projects/D...  8.731217e-07   
...                                                 ...           ...   
3195  C:/Users/Dylan/Desktop/Data Science/Projects/D...  3.715247e-10   
3196  C:/Users/Dylan/Desktop/Data Science/Projects/D...  1.574739e-06   
3197  C:/Users/Dylan/Desktop/Data Science/Projects/D...  1.789286e-10   
3198  C:/Users/Dylan/Desktop/Data Science/Projects/D...  6.973295e-01   
3199  C:/Users/Dylan/Desktop/Data Science/Projects/D...  4.641474e-01   

                c1            c2            c3            c4            c5  \
0     9.080641e-03  3.276221e-02  9.708376e-04  3.149434e-03  8.852060e-02   
1     9.999266e-01  2.778533e-09  2.121129e-05  9.715735e-09  8.799989e-11   
2     6.653389e-04  6.193582e-03  3.859256e-05  5.127674e-04  9.498447e-03   
3     9.999948e-01  2.486746e-11  2.996047e-07  3.825119e-11  7.392742e-14   
4     3.443508e-07  9.989744e-01  6.951326e-11  8.678994e-08  5.064950e-07   
...            ...           ...           ...           ...           ...   
3195  5.769833e-12  6.908332e-29  1.000000e+00  5.136259e-13  3.894515e-23   
3196  4.375948e-10  6.848797e-11  7.733332e-07  9.999092e-01  8.031368e-09   
3197  3.245966e-06  9.168875e-06  2.270906e-12  9.345583e-07  1.655015e-10   
3198  2.495644e-03  8.154131e-04  3.410958e-03  6.379608e-03  1.482557e-02   
3199  1.881661e-03  7.842500e-04  7.491834e-04  1.796803e-03  1.077689e-02   

                c6            c7            c8            c9  
0     1.004019e-02  4.595437e-01  2.119597e-01  1.168326e-01  
1     3.782401e-05  7.553241e-09  2.146324e-06  2.380554e-06  
2     2.100115e-03  8.335724e-01  1.233497e-01  1.939294e-02  
3     4.569526e-06  2.724793e-11  6.434313e-08  3.570090e-08  
4     7.010580e-05  1.246347e-04  8.061368e-04  2.294997e-05  
...            ...           ...           ...           ...  
3195  6.313904e-23  6.123699e-26  1.471512e-17  3.946067e-14  
3196  9.146721e-10  4.092025e-10  3.493802e-06  8.497512e-05  
3197  9.876809e-01  3.072587e-09  1.230577e-02  3.856044e-08  
3198  4.676780e-04  1.062956e-02  5.146125e-02  2.121847e-01  
3199  3.011495e-04  3.256380e-02  4.307273e-02  4.439261e-01  

[3200 rows x 14 columns]

In [19]:
# Save validation predictions
validation_predictions.to_csv('C:/Users/Dylan/Desktop/Data Science/Projects/DistractedDrivers/data/validation_predictions/model_1_validation_predictions.csv')

In [71]:
# Evaluate validation data with best coefficients
model_1_val_metrics = model_1.evaluate(val)

  ...
    to  
  ['...']
100/100 [==============================] - 20s 201ms/step - loss: 0.9765 - accuracy: 0.6913


In [73]:
model_1_val_metrics

[0.9764609852433205, 0.69125]

0.9765 cross-entropy loss and 0.6913 accuracy are not a bad start

### Model 2 - Xception Transfer Learning

In [26]:
model2 = tf.keras.applications.Xception(include_top=True,
                                        weights=None,
                                        input_shape=(227,227,3),
                                        classes=10)

In [27]:
model2.summary()

Model: "xception"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_3 (InputLayer)            [(None, 227, 227, 3) 0                                            
__________________________________________________________________________________________________
block1_conv1 (Conv2D)           (None, 113, 113, 32) 864         input_3[0][0]                    
__________________________________________________________________________________________________
block1_conv1_bn (BatchNormaliza (None, 113, 113, 32) 128         block1_conv1[0][0]               
__________________________________________________________________________________________________
block1_conv1_act (Activation)   (None, 113, 113, 32) 0           block1_conv1_bn[0][0]            
___________________________________________________________________________________________

In [28]:
opt = Adam(learning_rate=0.0001)

In [29]:
model2.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

In [30]:
model2.fit(train,
                epochs=50,
                steps_per_epoch=110,
                validation_data=test,
                callbacks=callbacks_list)

  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train for 110 steps, validate for 100 steps
Epoch 1/50
110/110 [==============================] - 77s 699ms/step - loss: 2.3130 - accuracy: 0.1250 - val_loss: 2.3027 - val_accuracy: 0.1000
Epoch 2/50
110/110 [==============================] - 63s 574ms/step - loss: 2.1135 - accuracy: 0.2233 - val_loss: 2.3029 - val_accuracy: 0.1197
Epoch 3/50
110/110 [==============================] - 65s 590ms/step - loss: 1.7786 - accuracy: 0.3409 - val_loss: 2.3041 - val_accuracy: 0.1000
Epoch 4/50
110/110 [==============================] - 64s 585ms/step - loss: 1.4025 - accuracy: 0.5017 - val_loss: 2.3077 - val_accuracy: 0.1000
Epoch 5/50
110/110 [==============================] - 63s 577ms/step - loss: 1.1029 - accuracy: 0.6114 - val_loss: 2.3216 - val_accuracy: 0.1000
Epoch 6/50
110/110 [==============================] - 67s 612ms/step - loss: 0.9071 - accuracy: 0.6898 - val_loss: 2.1428 - val_accuracy: 0.1575
Epoch 7/50
110/110 [================

In [31]:
validation_predictions = model2.predict(test)

In [32]:
validation_predictions = pd.DataFrame(validation_predictions, 
                                         columns=['c0', 'c1', 'c2', 'c3', 'c4', 'c5', 'c6', 'c7', 'c8', 'c9'])

In [33]:
classifier_train_data = df_test.reset_index(drop=True).join(validation_predictions)

In [34]:
classifier_train_data

subject classname            img                      imgpath        c0  \
0       p050        c9  img_37057.jpg  imgs/train/c9/img_37057.jpg  0.906768   
1       p056        c1  img_70214.jpg  imgs/train/c1/img_70214.jpg  0.000448   
2       p050        c7  img_89317.jpg  imgs/train/c7/img_89317.jpg  0.412319   
3       p041        c1   img_9714.jpg   imgs/train/c1/img_9714.jpg  0.076487   
4       p016        c2  img_46158.jpg  imgs/train/c2/img_46158.jpg  0.000089   
...      ...       ...            ...                          ...       ...   
3195    p041        c3  img_78650.jpg  imgs/train/c3/img_78650.jpg  0.029038   
3196    p041        c4  img_83455.jpg  imgs/train/c4/img_83455.jpg  0.013309   
3197    p041        c6  img_58453.jpg  imgs/train/c6/img_58453.jpg  0.000094   
3198    p041        c0  img_51874.jpg  imgs/train/c0/img_51874.jpg  0.578552   
3199    p056        c9  img_90247.jpg  imgs/train/c9/img_90247.jpg  0.678568   

            c1            c2            c3        c4        c5            c6  \
0     0.000744  7.208122e-05  2.896823e-03  0.000302  0.005444  4.678670e-05   
1     0.998688  1.731675e-07  1.224909e-04  0.000002  0.000002  7.495763e-05   
2     0.012335  3.431899e-02  3.826816e-02  0.001430  0.151051  5.680471e-04   
3     0.788871  3.576593e-05  1.271004e-05  0.004032  0.000011  9.147306e-04   
4     0.000009  9.915964e-01  3.862327e-07  0.000024  0.000110  4.323571e-04   
...        ...           ...           ...       ...       ...           ...   
3195  0.000017  4.017755e-07  9.692118e-01  0.000059  0.000465  1.427181e-06   
3196  0.000025  2.072875e-06  7.284578e-07  0.973240  0.000002  3.533776e-04   
3197  0.000005  7.073140e-05  1.910254e-06  0.013878  0.000003  9.320605e-01   
3198  0.000109  1.796256e-03  1.547626e-04  0.027229  0.003308  4.323568e-03   
3199  0.000696  2.363864e-07  7.556613e-05  0.000017  0.000036  6.784299e-07   

            c7        c8        c9  
0     0.001455  0.055513  0.026758  
1     0.000046  0.000157  0.000460  
2     0.185298  0.018996  0.145416  
3     0.000586  0.008436  0.120614  
4     0.000327  0.006047  0.001366  
...        ...       ...       ...  
3195  0.000048  0.000118  0.001042  
3196  0.000004  0.001772  0.011292  
3197  0.000038  0.053742  0.000107  
3198  0.013788  0.244803  0.125937  
3199  0.000070  0.000554  0.319982  

[3200 rows x 14 columns]

In [35]:
classifier_train_data.to_csv('../processed/Xception_classifier_training_data.csv')

In [36]:
from sklearn.model_selection import train_test_split

In [37]:
X = classifier_train_data.drop(columns=['subject', 'classname', 'img', 'imgpath'])
y= classifier_train_data[['classname']]

In [38]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

In [39]:
from sklearn.model_selection import RandomizedSearchCV
from sklearn.ensemble import RandomForestClassifier

In [40]:
params = {'criterion':['gini', 'entropy'],
          'min_samples_split':np.arange(2, 20, 1),
          'min_samples_leaf':np.arange(1, 20, 1)}

In [41]:
clf = RandomForestClassifier(n_jobs=-2)

In [42]:
model = RandomizedSearchCV(clf, params, scoring='accuracy', 
                           n_iter=30)

In [43]:
model.fit(X_train, np.ravel(y_train))

RandomizedSearchCV(estimator=RandomForestClassifier(n_jobs=-2), n_iter=30,
                   param_distributions={'criterion': ['gini', 'entropy'],
                                        'min_samples_leaf': array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19]),
                                        'min_samples_split': array([ 2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
       19])},
                   scoring='accuracy')

In [44]:
print(model.best_params_)

{'min_samples_split': 8, 'min_samples_leaf': 1, 'criterion': 'entropy'}


In [45]:
from sklearn.metrics import classification_report

In [46]:
y_pred = model.predict(X_test)

In [47]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

          c0       0.91      0.92      0.92        77
          c1       0.95      1.00      0.98        82
          c2       1.00      0.96      0.98        78
          c3       0.99      0.95      0.97        87
          c4       1.00      0.97      0.98        89
          c5       0.96      0.99      0.97        76
          c6       0.95      0.94      0.94        78
          c7       0.95      0.99      0.97        70
          c8       0.92      0.87      0.90        84
          c9       0.92      0.97      0.94        79

    accuracy                           0.95       800
   macro avg       0.95      0.96      0.95       800
weighted avg       0.96      0.95      0.95       800



Both models are created now. Let's evaluate on the small labeled test set.

In [48]:
os.getcwd()

'C:\\Users\\Dylan\\Desktop\\Data Science\\Projects\\DistractedDrivers\\data\\raw'

In [49]:
test_labeled_df = pd.read_csv('../processed/labeled_test_df.csv')

In [50]:
test_labeled_df

filename classname
0    img_101570.jpg        c0
1    img_102035.jpg        c0
2        img_22.jpg        c0
3        img_23.jpg        c0
4        img_24.jpg        c0
..              ...       ...
195   img_54733.jpg        c9
196   img_54928.jpg        c9
197   img_54990.jpg        c9
198   img_55171.jpg        c9
199   img_55896.jpg        c9

[200 rows x 2 columns]

In [51]:
os.getcwd()

'C:\\Users\\Dylan\\Desktop\\Data Science\\Projects\\DistractedDrivers\\data\\raw'

In [52]:
os.chdir('imgs/testlabeled/unsorted')

In [53]:
test = test_dgen.flow_from_dataframe(test_labeled_df,
                                     x_col='filename',
                                     y_col='classname',
                                     target_size=(227,227),
                                     batch_size=16,
                                     shuffle=False)

Found 200 validated image filenames belonging to 10 classes.


In [54]:
test_nn_output = model2.predict(test)

In [55]:
model2.evaluate(test)

  ...
    to  
  ['...']
13/13 [==============================] - 1s 102ms/step - loss: 1.1134 - accuracy: 0.6900


[1.1133543573893034, 0.69]

In [56]:
y_labeled_test=test_labeled_df[['classname']]

In [57]:
X_labeled_test = pd.DataFrame(test_nn_output, 
                              columns=['c0', 'c1', 'c2', 'c3', 'c4', 'c5', 'c6', 'c7', 'c8', 'c9'])

In [58]:
X_labeled_test

c0            c1            c2            c3        c4        c5  \
0    0.867632  1.209877e-04  1.500739e-03  2.442012e-04  0.000436  0.030791   
1    0.085785  9.925861e-08  4.517725e-06  1.069515e-06  0.000032  0.013126   
2    0.939698  1.831112e-05  1.414316e-06  4.492516e-04  0.000031  0.001389   
3    0.775485  5.481432e-05  1.470785e-07  2.721378e-05  0.000013  0.000008   
4    0.853957  4.373438e-06  3.896049e-08  4.490081e-04  0.000052  0.000003   
..        ...           ...           ...           ...       ...       ...   
195  0.065003  8.777229e-05  6.820670e-05  4.324991e-04  0.000064  0.047161   
196  0.006278  7.241461e-05  1.034660e-04  9.460746e-07  0.000542  0.025136   
197  0.480909  1.950020e-05  1.225575e-06  1.957746e-05  0.000031  0.000019   
198  0.025519  6.326396e-04  9.093390e-07  4.500289e-05  0.000056  0.000069   
199  0.292451  1.541344e-04  2.597889e-05  5.561590e-05  0.003590  0.001364   

               c6        c7        c8        c9  
0    9.351860e-05  0.010375  0.010932  0.077874  
1    8.762860e-06  0.000142  0.498029  0.402872  
2    3.769517e-06  0.000115  0.007601  0.050695  
3    3.541720e-06  0.000037  0.022309  0.202061  
4    7.822570e-08  0.000009  0.001505  0.144020  
..            ...       ...       ...       ...  
195  2.973797e-05  0.000613  0.001563  0.884978  
196  1.378974e-03  0.000038  0.205333  0.761117  
197  1.114335e-06  0.000024  0.004597  0.514378  
198  1.127034e-05  0.219518  0.027573  0.726576  
199  6.599955e-04  0.000792  0.223005  0.477901  

[200 rows x 10 columns]

In [59]:
y_labeled_test

classname
0          c0
1          c0
2          c0
3          c0
4          c0
..        ...
195        c9
196        c9
197        c9
198        c9
199        c9

[200 rows x 1 columns]

In [60]:
labeled_predictions = model.predict(X_labeled_test)

In [61]:
labeled_predictions

array(['c0', 'c9', 'c0', 'c9', 'c0', 'c0', 'c0', 'c8', 'c0', 'c0', 'c0',
       'c0', 'c0', 'c9', 'c0', 'c9', 'c0', 'c0', 'c5', 'c0', 'c1', 'c1',
       'c1', 'c9', 'c6', 'c6', 'c1', 'c1', 'c5', 'c2', 'c9', 'c1', 'c0',
       'c1', 'c1', 'c1', 'c6', 'c6', 'c3', 'c1', 'c2', 'c2', 'c2', 'c2',
       'c8', 'c2', 'c2', 'c2', 'c2', 'c2', 'c2', 'c8', 'c2', 'c9', 'c2',
       'c2', 'c2', 'c2', 'c2', 'c2', 'c3', 'c3', 'c3', 'c3', 'c3', 'c3',
       'c3', 'c3', 'c3', 'c3', 'c3', 'c3', 'c3', 'c3', 'c3', 'c3', 'c5',
       'c3', 'c0', 'c3', 'c0', 'c4', 'c0', 'c4', 'c0', 'c4', 'c4', 'c4',
       'c4', 'c4', 'c0', 'c0', 'c4', 'c4', 'c4', 'c9', 'c3', 'c4', 'c4',
       'c4', 'c5', 'c5', 'c5', 'c5', 'c5', 'c5', 'c5', 'c5', 'c5', 'c5',
       'c5', 'c5', 'c5', 'c5', 'c9', 'c5', 'c5', 'c5', 'c5', 'c5', 'c8',
       'c6', 'c6', 'c8', 'c6', 'c6', 'c8', 'c6', 'c6', 'c6', 'c8', 'c8',
       'c8', 'c6', 'c8', 'c0', 'c6', 'c6', 'c6', 'c6', 'c8', 'c8', 'c8',
       'c8', 'c8', 'c8', 'c8', 'c8', 'c8', 'c0', 'c

In [62]:
print(classification_report(y_labeled_test, labeled_predictions))

              precision    recall  f1-score   support

          c0       0.52      0.70      0.60        20
          c1       1.00      0.50      0.67        20
          c2       0.89      0.85      0.87        20
          c3       0.90      0.90      0.90        20
          c4       1.00      0.65      0.79        20
          c5       0.76      0.95      0.84        20
          c6       0.71      0.60      0.65        20
          c7       0.67      0.10      0.17        20
          c8       0.38      0.75      0.50        20
          c9       0.54      0.70      0.61        20

    accuracy                           0.67       200
   macro avg       0.74      0.67      0.66       200
weighted avg       0.74      0.67      0.66       200



In [63]:
print(model.predict_proba(X_labeled_test))

[[0.29688889 0.         0.001      ... 0.21879816 0.06232143 0.18760317]
 [0.06813492 0.02       0.06015873 ... 0.03075    0.16754762 0.34258694]
 [0.7086627  0.         0.         ... 0.001      0.00666667 0.27128968]
 ...
 [0.09308081 0.         0.         ... 0.         0.         0.87284343]
 [0.09182179 0.02690909 0.005      ... 0.33226496 0.20396087 0.24120202]
 [0.59651587 0.         0.08571032 ... 0.01966667 0.06280159 0.11502778]]


### VGG16 Architecture - Transfer-Learning

Using VGG16 with imagenet weights as base model and then training final output layers

In [62]:
base_model3 = tf.keras.applications.VGG16(include_top=False,
                                        weights='imagenet',
                                        input_shape=(227,227,3))

In [63]:
for layer in model3.layers:
    layer.trainable = False

In [64]:
from tensorflow.keras import layers

In [65]:
# Flatten the output layer to 1 dimension
x = layers.Flatten()(model3.output)

# Add a fully connected layer with 512 hidden units and ReLU activation
x = layers.Dense(512, activation='relu')(x)

# Add a dropout rate of 0.5
x = layers.Dropout(0.5)(x)

# Add a final sigmoid layer for classification
x = layers.Dense(10, activation='softmax')(x)

model3 = tf.keras.models.Model(model3.input, x)

model3.compile(optimizer = tf.keras.optimizers.RMSprop(lr=0.0001), loss = 'categorical_crossentropy',metrics = ['acc'])

In [66]:
model3.summary()

Model: "model_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_3 (InputLayer)         [(None, 227, 227, 3)]     0         
_________________________________________________________________
block1_conv1 (Conv2D)        (None, 227, 227, 64)      1792      
_________________________________________________________________
block1_conv2 (Conv2D)        (None, 227, 227, 64)      36928     
_________________________________________________________________
block1_pool (MaxPooling2D)   (None, 113, 113, 64)      0         
_________________________________________________________________
block2_conv1 (Conv2D)        (None, 113, 113, 128)     73856     
_________________________________________________________________
block2_conv2 (Conv2D)        (None, 113, 113, 128)     147584    
_________________________________________________________________
block2_pool (MaxPooling2D)   (None, 56, 56, 128)       0   

In [67]:
os.getcwd()

'C:\\Users\\Dylan\\Desktop\\Data Science\\Projects\\DistractedDrivers\\data\\raw\\imgs\\testlabeled\\unsorted'

In [70]:
model3.load_weights('C:\\Users\\Dylan\\Desktop\\Data Science\\Projects\\DistractedDrivers\\data\\raw\\weights.hdf5')

In [33]:
model3.fit(train,
                epochs=50,
                steps_per_epoch=110,
                validation_data=test,
                callbacks=callbacks_list)

  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train for 110 steps, validate for 100 steps
Epoch 1/50
110/110 [==============================] - 73s 667ms/step - loss: 2.5233 - acc: 0.1318 - val_loss: 2.2212 - val_acc: 0.1472
Epoch 2/50
110/110 [==============================] - 73s 666ms/step - loss: 2.2779 - acc: 0.1562 - val_loss: 2.1070 - val_acc: 0.3166
Epoch 3/50
110/110 [==============================] - 73s 663ms/step - loss: 2.1991 - acc: 0.2000 - val_loss: 1.9553 - val_acc: 0.4006
Epoch 4/50
110/110 [==============================] - 68s 621ms/step - loss: 2.0999 - acc: 0.2307 - val_loss: 1.8996 - val_acc: 0.4181
Epoch 5/50
110/110 [==============================] - 71s 644ms/step - loss: 2.0020 - acc: 0.2710 - val_loss: 1.6876 - val_acc: 0.4712
Epoch 6/50
110/110 [==============================] - 65s 587ms/step - loss: 1.9430 - acc: 0.3136 - val_loss: 1.6530 - val_acc: 0.5459
Epoch 7/50
110/110 [==============================] - 65s 590ms/step - loss: 1.8607 - acc: 0.343

In [34]:
validation_predictions = model3.predict(test)

In [35]:
validation_predictions = pd.DataFrame(validation_predictions, 
                                         columns=['c0', 'c1', 'c2', 'c3', 'c4', 'c5', 'c6', 'c7', 'c8', 'c9'])

In [36]:
classifier_train_data = df_test.reset_index(drop=True).join(validation_predictions)

In [37]:
classifier_train_data

subject classname            img                      imgpath        c0  \
0       p050        c9  img_37057.jpg  imgs/train/c9/img_37057.jpg  0.038092   
1       p056        c1  img_70214.jpg  imgs/train/c1/img_70214.jpg  0.001814   
2       p050        c7  img_89317.jpg  imgs/train/c7/img_89317.jpg  0.004418   
3       p041        c1   img_9714.jpg   imgs/train/c1/img_9714.jpg  0.021931   
4       p016        c2  img_46158.jpg  imgs/train/c2/img_46158.jpg  0.003152   
...      ...       ...            ...                          ...       ...   
3195    p041        c3  img_78650.jpg  imgs/train/c3/img_78650.jpg  0.009461   
3196    p041        c4  img_83455.jpg  imgs/train/c4/img_83455.jpg  0.007229   
3197    p041        c6  img_58453.jpg  imgs/train/c6/img_58453.jpg  0.005851   
3198    p041        c0  img_51874.jpg  imgs/train/c0/img_51874.jpg  0.409633   
3199    p056        c9  img_90247.jpg  imgs/train/c9/img_90247.jpg  0.174123   

            c1        c2        c3        c4        c5        c6        c7  \
0     0.047378  0.055674  0.003527  0.011132  0.032342  0.036965  0.002194   
1     0.665885  0.046201  0.011587  0.001652  0.002267  0.147198  0.049998   
2     0.058346  0.321970  0.004580  0.001134  0.001784  0.017168  0.522277   
3     0.579883  0.051502  0.050977  0.004998  0.064224  0.037028  0.003008   
4     0.329363  0.136421  0.014852  0.012734  0.014436  0.121248  0.182954   
...        ...       ...       ...       ...       ...       ...       ...   
3195  0.001977  0.004556  0.822273  0.130280  0.014533  0.001842  0.000328   
3196  0.000311  0.011610  0.164005  0.793564  0.010024  0.003079  0.000376   
3197  0.024534  0.390130  0.028171  0.084743  0.035849  0.285895  0.004593   
3198  0.021191  0.047486  0.014772  0.004590  0.097627  0.006294  0.039794   
3199  0.042037  0.019459  0.026641  0.004215  0.016195  0.003116  0.078980   

            c8        c9  
0     0.729606  0.043091  
1     0.066739  0.006660  
2     0.043075  0.025248  
3     0.048320  0.138128  
4     0.082555  0.102285  
...        ...       ...  
3195  0.008226  0.006525  
3196  0.006493  0.003310  
3197  0.124124  0.016109  
3198  0.050137  0.308477  
3199  0.019172  0.616062  

[3200 rows x 14 columns]

In [38]:
classifier_train_data.to_csv('../processed/VGG16_classifier_training_data.csv')

In [39]:
from sklearn.model_selection import train_test_split

In [40]:
X = classifier_train_data.drop(columns=['subject', 'classname', 'img', 'imgpath'])
y= classifier_train_data[['classname']]

In [41]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

In [42]:
from sklearn.model_selection import RandomizedSearchCV
from sklearn.ensemble import RandomForestClassifier

In [43]:
params = {'criterion':['gini', 'entropy'],
          'min_samples_split':np.arange(2, 20, 1),
          'min_samples_leaf':np.arange(1, 20, 1)}

In [71]:
clf = RandomForestClassifier(n_jobs=-2)

In [74]:
RandForest = RandomizedSearchCV(clf, params, scoring='accuracy', 
                           n_iter=30)

In [75]:
RandForest.fit(X_train, np.ravel(y_train))

RandomizedSearchCV(estimator=RandomForestClassifier(n_jobs=-2), n_iter=30,
                   param_distributions={'criterion': ['gini', 'entropy'],
                                        'min_samples_leaf': array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19]),
                                        'min_samples_split': array([ 2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
       19])},
                   scoring='accuracy')

In [76]:
print(RandForest.best_params_)

{'min_samples_split': 5, 'min_samples_leaf': 2, 'criterion': 'entropy'}


In [77]:
from sklearn.metrics import classification_report

In [78]:
y_pred = RandForest.predict(X_test)

In [79]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

          c0       0.88      0.97      0.93        77
          c1       0.96      0.96      0.96        82
          c2       0.92      0.92      0.92        78
          c3       0.99      0.95      0.97        87
          c4       0.98      0.99      0.98        89
          c5       1.00      1.00      1.00        76
          c6       0.89      0.95      0.92        78
          c7       0.95      0.99      0.97        70
          c8       0.99      0.85      0.91        84
          c9       0.97      0.95      0.96        79

    accuracy                           0.95       800
   macro avg       0.95      0.95      0.95       800
weighted avg       0.95      0.95      0.95       800



Both models are created now. Let's evaluate on the small labeled test set.

In [55]:
os.getcwd()

'C:\\Users\\Dylan\\Desktop\\Data Science\\Projects\\DistractedDrivers\\data\\raw'

In [56]:
test_labeled_df = pd.read_csv('../processed/labeled_test_df.csv')

In [80]:
test_labeled_df

filename classname
0    img_101570.jpg        c0
1    img_102035.jpg        c0
2        img_22.jpg        c0
3        img_23.jpg        c0
4        img_24.jpg        c0
..              ...       ...
195   img_54733.jpg        c9
196   img_54928.jpg        c9
197   img_54990.jpg        c9
198   img_55171.jpg        c9
199   img_55896.jpg        c9

[200 rows x 2 columns]

In [81]:
os.getcwd()

'C:\\Users\\Dylan\\Desktop\\Data Science\\Projects\\DistractedDrivers\\data\\raw\\imgs\\testlabeled\\unsorted'

In [82]:
os.chdir('imgs/testlabeled/unsorted')

FileNotFoundError: [WinError 3] The system cannot find the path specified: 'imgs/testlabeled/unsorted'

In [83]:
test = test_dgen.flow_from_dataframe(test_labeled_df,
                                     x_col='filename',
                                     y_col='classname',
                                     target_size=(227,227),
                                     batch_size=16,
                                     shuffle=False)

Found 200 validated image filenames belonging to 10 classes.


In [84]:
test_nn_output = model3.predict(test)

In [85]:
model3.evaluate(test)

  ...
    to  
  ['...']
13/13 [==============================] - 2s 125ms/step - loss: 0.8906 - acc: 0.7300


[0.8905925131761111, 0.73]

In [86]:
y_labeled_test=test_labeled_df[['classname']]

In [87]:
X_labeled_test = pd.DataFrame(test_nn_output, 
                              columns=['c0', 'c1', 'c2', 'c3', 'c4', 'c5', 'c6', 'c7', 'c8', 'c9'])

In [88]:
X_labeled_test

c0        c1        c2        c3        c4        c5        c6  \
0    0.047275  0.005121  0.027903  0.069816  0.096730  0.443145  0.157516   
1    0.757133  0.000159  0.000689  0.019632  0.008444  0.082580  0.000070   
2    0.393718  0.007405  0.000747  0.249553  0.039259  0.031297  0.001105   
3    0.469957  0.009404  0.002149  0.127384  0.066894  0.024054  0.001107   
4    0.102190  0.102216  0.010231  0.464914  0.107785  0.008216  0.053829   
..        ...       ...       ...       ...       ...       ...       ...   
195  0.004901  0.107628  0.026312  0.000329  0.000066  0.002584  0.002301   
196  0.001136  0.006047  0.019866  0.002046  0.004147  0.001085  0.005265   
197  0.486174  0.046821  0.000994  0.012781  0.000907  0.115279  0.001817   
198  0.106881  0.007360  0.058115  0.084603  0.080829  0.007823  0.006192   
199  0.069602  0.038255  0.028288  0.071635  0.063002  0.039964  0.006913   

           c7        c8        c9  
0    0.006203  0.106239  0.040052  
1    0.000231  0.001939  0.129122  
2    0.001983  0.012218  0.262715  
3    0.005694  0.015252  0.278104  
4    0.014488  0.033943  0.102189  
..        ...       ...       ...  
195  0.498552  0.013659  0.343667  
196  0.003656  0.944786  0.011967  
197  0.001643  0.005545  0.328040  
198  0.143699  0.163061  0.341437  
199  0.008187  0.057093  0.617061  

[200 rows x 10 columns]

In [89]:
y_labeled_test

classname
0          c0
1          c0
2          c0
3          c0
4          c0
..        ...
195        c9
196        c9
197        c9
198        c9
199        c9

[200 rows x 1 columns]

In [94]:
labeled_predictions = RandForest.predict(X_labeled_test)

In [95]:
labeled_predictions

array(['c6', 'c0', 'c0', 'c0', 'c3', 'c0', 'c0', 'c0', 'c0', 'c0', 'c5',
       'c0', 'c0', 'c0', 'c0', 'c4', 'c0', 'c0', 'c0', 'c0', 'c1', 'c1',
       'c1', 'c1', 'c1', 'c8', 'c2', 'c1', 'c1', 'c6', 'c1', 'c8', 'c1',
       'c1', 'c6', 'c1', 'c1', 'c2', 'c1', 'c1', 'c6', 'c6', 'c8', 'c2',
       'c1', 'c7', 'c1', 'c1', 'c2', 'c6', 'c7', 'c2', 'c2', 'c8', 'c1',
       'c7', 'c8', 'c2', 'c2', 'c2', 'c3', 'c5', 'c3', 'c4', 'c3', 'c3',
       'c3', 'c3', 'c4', 'c4', 'c3', 'c3', 'c3', 'c3', 'c3', 'c3', 'c3',
       'c3', 'c0', 'c3', 'c8', 'c3', 'c4', 'c4', 'c8', 'c4', 'c4', 'c4',
       'c4', 'c4', 'c3', 'c3', 'c4', 'c4', 'c4', 'c4', 'c4', 'c3', 'c4',
       'c4', 'c5', 'c5', 'c5', 'c5', 'c5', 'c5', 'c5', 'c5', 'c5', 'c5',
       'c5', 'c5', 'c5', 'c5', 'c8', 'c5', 'c5', 'c5', 'c5', 'c5', 'c6',
       'c6', 'c6', 'c6', 'c6', 'c1', 'c6', 'c6', 'c6', 'c6', 'c6', 'c1',
       'c8', 'c8', 'c6', 'c8', 'c6', 'c6', 'c6', 'c6', 'c7', 'c7', 'c7',
       'c8', 'c8', 'c8', 'c7', 'c8', 'c1', 'c7', 'c

In [96]:
print(classification_report(y_labeled_test, labeled_predictions))

              precision    recall  f1-score   support

          c0       0.70      0.80      0.74        20
          c1       0.61      0.70      0.65        20
          c2       0.78      0.35      0.48        20
          c3       0.71      0.75      0.73        20
          c4       0.78      0.70      0.74        20
          c5       0.86      0.95      0.90        20
          c6       0.65      0.75      0.70        20
          c7       0.73      0.55      0.63        20
          c8       0.42      0.80      0.55        20
          c9       1.00      0.40      0.57        20

    accuracy                           0.68       200
   macro avg       0.72      0.68      0.67       200
weighted avg       0.72      0.68      0.67       200



In [97]:
print(RandForest.predict_proba(X_labeled_test))

[[0.01       0.006      0.04404762 ... 0.002      0.1195     0.01      ]
 [0.90940476 0.         0.         ... 0.01       0.04083333 0.0225    ]
 [0.94166667 0.         0.         ... 0.         0.         0.05833333]
 ...
 [0.92666667 0.         0.         ... 0.         0.         0.06      ]
 [0.2202619  0.00666667 0.0205     ... 0.015      0.17980952 0.36592857]
 [0.11433333 0.00916667 0.0165     ... 0.         0.022      0.6155    ]]


Load datagen for actual test data

In [49]:
test_filenames = os.listdir('data/raw/imgs/test/test')

In [50]:
test_filenames = pd.DataFrame({'filename':test_filenames})

In [51]:
test_filenames

filename
0           img_1.jpg
1          img_10.jpg
2         img_100.jpg
3        img_1000.jpg
4      img_100000.jpg
...               ...
79721   img_99994.jpg
79722   img_99995.jpg
79723   img_99996.jpg
79724   img_99998.jpg
79725   img_99999.jpg

[79726 rows x 1 columns]

In [52]:
test_filenames.to_csv('data/raw/test_filenames.csv', index=False)

In [59]:
df_test

filename
0           img_1.jpg
1          img_10.jpg
2         img_100.jpg
3        img_1000.jpg
4      img_100000.jpg
...               ...
79721   img_99994.jpg
79722   img_99995.jpg
79723   img_99996.jpg
79724   img_99998.jpg
79725   img_99999.jpg

[79726 rows x 1 columns]

In [60]:
df_test['filename'] = df_test[['filename']].apply(lambda x: 'C:/Users/Dylan/Desktop/Data Science/Projects/DistractedDrivers/data/raw/imgs/test/test/' + x, axis=1)

In [61]:
df_test

filename
0      C:/Users/Dylan/Desktop/Data Science/Projects/D...
1      C:/Users/Dylan/Desktop/Data Science/Projects/D...
2      C:/Users/Dylan/Desktop/Data Science/Projects/D...
3      C:/Users/Dylan/Desktop/Data Science/Projects/D...
4      C:/Users/Dylan/Desktop/Data Science/Projects/D...
...                                                  ...
79721  C:/Users/Dylan/Desktop/Data Science/Projects/D...
79722  C:/Users/Dylan/Desktop/Data Science/Projects/D...
79723  C:/Users/Dylan/Desktop/Data Science/Projects/D...
79724  C:/Users/Dylan/Desktop/Data Science/Projects/D...
79725  C:/Users/Dylan/Desktop/Data Science/Projects/D...

[79726 rows x 1 columns]

In [105]:
os.getcwd()

'C:\\Users\\Dylan\\Desktop\\Data Science\\Projects\\DistractedDrivers\\data\\raw\\imgs\\test\\test'

In [62]:
df_test['class'] = 'unknown'

In [64]:
df_test.to_csv('C:/Users/Dylan/Desktop/Data Science/Projects/DistractedDrivers/data/raw/test_filenames.csv', index=False)

In [107]:
test = test_dgen.flow_from_dataframe(test_filenames,
                                     target_size=(227,227),
                                     batch_size=16,
                                     shuffle=False)

Found 79726 validated image filenames belonging to 1 classes.


In [108]:
test_predictions = model3.predict(test)

In [109]:
test_predictions = pd.DataFrame(test_predictions, columns=['c0', 'c1', 'c2', 'c3', 'c4', 'c5', 'c6', 'c7', 'c8', 'c9'])

In [110]:
test_predictions

c0        c1        c2        c3        c4        c5        c6  \
0      0.008604  0.085207  0.018095  0.113947  0.040185  0.600935  0.077657   
1      0.019094  0.000826  0.001663  0.006799  0.003642  0.941262  0.001846   
2      0.200818  0.249065  0.058669  0.010818  0.004579  0.048115  0.014242   
3      0.004387  0.002688  0.024336  0.000805  0.002280  0.009872  0.004561   
4      0.045187  0.001458  0.000840  0.701084  0.199329  0.022523  0.005878   
...         ...       ...       ...       ...       ...       ...       ...   
79721  0.004477  0.186477  0.133851  0.001749  0.000536  0.011587  0.029475   
79722  0.013295  0.035097  0.000265  0.931134  0.004813  0.001793  0.002391   
79723  0.000791  0.041994  0.000147  0.846906  0.099453  0.000325  0.007390   
79724  0.000449  0.085182  0.286448  0.000522  0.000433  0.000592  0.328923   
79725  0.036727  0.002992  0.000412  0.005636  0.001364  0.934877  0.000635   

             c7        c8        c9  
0      0.006483  0.026404  0.022482  
1      0.000832  0.017235  0.006800  
2      0.062749  0.039698  0.311247  
3      0.000767  0.943617  0.006687  
4      0.001054  0.016436  0.006211  
...         ...       ...       ...  
79721  0.449997  0.096771  0.085079  
79722  0.000746  0.002524  0.007941  
79723  0.000138  0.002115  0.000740  
79724  0.205457  0.084118  0.007875  
79725  0.000332  0.008541  0.008485  

[79726 rows x 10 columns]

In [121]:
test_predictions_df = test_filenames.join(test_predictions)
test_predictions_df.drop(columns=['class'], inplace=True)
test_predictions_df.rename(columns={'filename':'img'}, inplace=True)
test_predictions_df

img        c0        c1        c2        c3        c4  \
0           img_1.jpg  0.008604  0.085207  0.018095  0.113947  0.040185   
1          img_10.jpg  0.019094  0.000826  0.001663  0.006799  0.003642   
2         img_100.jpg  0.200818  0.249065  0.058669  0.010818  0.004579   
3        img_1000.jpg  0.004387  0.002688  0.024336  0.000805  0.002280   
4      img_100000.jpg  0.045187  0.001458  0.000840  0.701084  0.199329   
...               ...       ...       ...       ...       ...       ...   
79721   img_99994.jpg  0.004477  0.186477  0.133851  0.001749  0.000536   
79722   img_99995.jpg  0.013295  0.035097  0.000265  0.931134  0.004813   
79723   img_99996.jpg  0.000791  0.041994  0.000147  0.846906  0.099453   
79724   img_99998.jpg  0.000449  0.085182  0.286448  0.000522  0.000433   
79725   img_99999.jpg  0.036727  0.002992  0.000412  0.005636  0.001364   

             c5        c6        c7        c8        c9  
0      0.600935  0.077657  0.006483  0.026404  0.022482  
1      0.941262  0.001846  0.000832  0.017235  0.006800  
2      0.048115  0.014242  0.062749  0.039698  0.311247  
3      0.009872  0.004561  0.000767  0.943617  0.006687  
4      0.022523  0.005878  0.001054  0.016436  0.006211  
...         ...       ...       ...       ...       ...  
79721  0.011587  0.029475  0.449997  0.096771  0.085079  
79722  0.001793  0.002391  0.000746  0.002524  0.007941  
79723  0.000325  0.007390  0.000138  0.002115  0.000740  
79724  0.000592  0.328923  0.205457  0.084118  0.007875  
79725  0.934877  0.000635  0.000332  0.008541  0.008485  

[79726 rows x 11 columns]

In [126]:
test_predictions_df.to_csv('C:\\Users\\Dylan\\Desktop\\Data Science\\Projects\\DistractedDrivers\\data\\test_predictions(VGG16_Transfer_Learning).csv', index=False)

In [111]:
rf_predictions = model.predict_proba(test_predictions)

In [112]:
rf_predictions

array([[0.01277273, 0.05188699, 0.06216909, ..., 0.0022028 , 0.07507054,
        0.02621848],
       [0.        , 0.        , 0.        , ..., 0.        , 0.0188447 ,
        0.00457516],
       [0.50327548, 0.19732263, 0.00961905, ..., 0.        , 0.02691071,
        0.22182527],
       ...,
       [0.00173611, 0.        , 0.        , ..., 0.        , 0.00291667,
        0.        ],
       [0.        , 0.        , 0.19470271, ..., 0.38539343, 0.19947546,
        0.03235011],
       [0.        , 0.        , 0.00482759, ..., 0.00206897, 0.0069697 ,
        0.01766667]])

In [113]:
rf_test_predictions = pd.DataFrame(rf_predictions, columns=['c0', 'c1', 'c2', 'c3', 'c4', 'c5', 'c6', 'c7', 'c8', 'c9'])

In [114]:
predictions = test_filenames.join(rf_test_predictions)

In [115]:
predictions.drop(columns=['class'], inplace=True)

In [116]:
predictions.rename(columns={'filename': 'img'}, inplace=True)

In [117]:
predictions

img        c0        c1        c2        c3        c4  \
0           img_1.jpg  0.012773  0.051887  0.062169  0.045604  0.075507   
1          img_10.jpg  0.000000  0.000000  0.000000  0.000000  0.011932   
2         img_100.jpg  0.503275  0.197323  0.009619  0.003040  0.000000   
3        img_1000.jpg  0.000526  0.001600  0.003887  0.003750  0.005671   
4      img_100000.jpg  0.006813  0.000000  0.000000  0.957205  0.029316   
...               ...       ...       ...       ...       ...       ...   
79721   img_99994.jpg  0.028851  0.258388  0.141071  0.000000  0.000000   
79722   img_99995.jpg  0.012290  0.010714  0.021952  0.854877  0.000000   
79723   img_99996.jpg  0.001736  0.000000  0.000000  0.991806  0.001667   
79724   img_99998.jpg  0.000000  0.000000  0.194703  0.000000  0.000000   
79725   img_99999.jpg  0.000000  0.000000  0.004828  0.000000  0.012487   

             c5        c6        c7        c8        c9  
0      0.612750  0.035819  0.002203  0.075071  0.026218  
1      0.964648  0.000000  0.000000  0.018845  0.004575  
2      0.024722  0.013285  0.000000  0.026911  0.221825  
3      0.037143  0.023430  0.010000  0.770966  0.143027  
4      0.001250  0.000000  0.000000  0.005417  0.000000  
...         ...       ...       ...       ...       ...  
79721  0.009987  0.022551  0.240912  0.170321  0.127919  
79722  0.039625  0.010625  0.022601  0.012248  0.015068  
79723  0.001250  0.000625  0.000000  0.002917  0.000000  
79724  0.001304  0.186774  0.385393  0.199475  0.032350  
79725  0.955980  0.000000  0.002069  0.006970  0.017667  

[79726 rows x 11 columns]

In [118]:
os.getcwd()

'C:\\Users\\Dylan\\Desktop\\Data Science\\Projects\\DistractedDrivers\\data\\raw\\imgs\\test\\test'

In [127]:
predictions.to_csv('C:\\Users\\Dylan\\Desktop\\Data Science\\Projects\\DistractedDrivers\\data\\test_predictions(VGG16_Transfer_Learning_with_RandFor).csv', index=False)

In [124]:
test_predictions_df

img        c0        c1        c2        c3        c4  \
0           img_1.jpg  0.008604  0.085207  0.018095  0.113947  0.040185   
1          img_10.jpg  0.019094  0.000826  0.001663  0.006799  0.003642   
2         img_100.jpg  0.200818  0.249065  0.058669  0.010818  0.004579   
3        img_1000.jpg  0.004387  0.002688  0.024336  0.000805  0.002280   
4      img_100000.jpg  0.045187  0.001458  0.000840  0.701084  0.199329   
...               ...       ...       ...       ...       ...       ...   
79721   img_99994.jpg  0.004477  0.186477  0.133851  0.001749  0.000536   
79722   img_99995.jpg  0.013295  0.035097  0.000265  0.931134  0.004813   
79723   img_99996.jpg  0.000791  0.041994  0.000147  0.846906  0.099453   
79724   img_99998.jpg  0.000449  0.085182  0.286448  0.000522  0.000433   
79725   img_99999.jpg  0.036727  0.002992  0.000412  0.005636  0.001364   

             c5        c6        c7        c8        c9  
0      0.600935  0.077657  0.006483  0.026404  0.022482  
1      0.941262  0.001846  0.000832  0.017235  0.006800  
2      0.048115  0.014242  0.062749  0.039698  0.311247  
3      0.009872  0.004561  0.000767  0.943617  0.006687  
4      0.022523  0.005878  0.001054  0.016436  0.006211  
...         ...       ...       ...       ...       ...  
79721  0.011587  0.029475  0.449997  0.096771  0.085079  
79722  0.001793  0.002391  0.000746  0.002524  0.007941  
79723  0.000325  0.007390  0.000138  0.002115  0.000740  
79724  0.000592  0.328923  0.205457  0.084118  0.007875  
79725  0.934877  0.000635  0.000332  0.008541  0.008485  

[79726 rows x 11 columns]

In [125]:
predictions

img        c0        c1        c2        c3        c4  \
0           img_1.jpg  0.012773  0.051887  0.062169  0.045604  0.075507   
1          img_10.jpg  0.000000  0.000000  0.000000  0.000000  0.011932   
2         img_100.jpg  0.503275  0.197323  0.009619  0.003040  0.000000   
3        img_1000.jpg  0.000526  0.001600  0.003887  0.003750  0.005671   
4      img_100000.jpg  0.006813  0.000000  0.000000  0.957205  0.029316   
...               ...       ...       ...       ...       ...       ...   
79721   img_99994.jpg  0.028851  0.258388  0.141071  0.000000  0.000000   
79722   img_99995.jpg  0.012290  0.010714  0.021952  0.854877  0.000000   
79723   img_99996.jpg  0.001736  0.000000  0.000000  0.991806  0.001667   
79724   img_99998.jpg  0.000000  0.000000  0.194703  0.000000  0.000000   
79725   img_99999.jpg  0.000000  0.000000  0.004828  0.000000  0.012487   

             c5        c6        c7        c8        c9  
0      0.612750  0.035819  0.002203  0.075071  0.026218  
1      0.964648  0.000000  0.000000  0.018845  0.004575  
2      0.024722  0.013285  0.000000  0.026911  0.221825  
3      0.037143  0.023430  0.010000  0.770966  0.143027  
4      0.001250  0.000000  0.000000  0.005417  0.000000  
...         ...       ...       ...       ...       ...  
79721  0.009987  0.022551  0.240912  0.170321  0.127919  
79722  0.039625  0.010625  0.022601  0.012248  0.015068  
79723  0.001250  0.000625  0.000000  0.002917  0.000000  
79724  0.001304  0.186774  0.385393  0.199475  0.032350  
79725  0.955980  0.000000  0.002069  0.006970  0.017667  

[79726 rows x 11 columns]